### PR TITLE
Change SqlClient to use strongly typed packet and session handles

### PR DESCRIPTION
**Pull request by Wraith2**
These changes remove constant boxing of session and packet handles which are IntPtr when using the native implementation of `TdsParserStateObject` (native is used in windows non-uap builds) which improves memory performance. Making these changes also allows assertion of the handle types in debug builds which adds safety and testability. 

To do this I have had to add in a project DefineConstant called `snidll` to allow some fields to be omitted where the native implementation is not included, while I like the current configuration with extremely minimal #define usage I considered this and think the additional type safety offered is worth the new constant. I have also renamed a couple of types to clarify which types are used in which implementations.

The changes also include some changes to the ManualTests to add conditional facts on Udt and ServiceBroker reliant test cases so they won't run if the resources required aren't present or enabled. This assumes that a full test run with these available should not skip any tests. These changes were needed to allow me to run the manual tests locally without getting large numbers of errors to try and pick real problems out of.

please review, test, comment, say "oh no not this guy again" as usual 😁 

/cc @keeratsingh @AfsanehR @saurabh500 @David-Engel

### CODE REVIEW

File: src/System.Data.SqlClient/src/System/Data/SqlClient/SNI/SNIPacket.cs [hunk 1]
```diff
@@ -98,14 +98,14 @@ public void Allocate(int capacity)
             {
                 if (_isBufferFromArrayPool)
                 {
-                    _arrayPool.Return(_data);
+                    ArrayPool<byte>.Shared.Return(_data);
```
[the remark this answers is not in the document]
Comment (benaadams) [reply src/System.Data.SqlClient/src/System/Data/SqlClient/SNI/SNIPacket.cs]: > The default pool can also be devirtualized by the jit possibly making the calls slightly faster.

Yep, when referred to directly via `ArrayPool<byte>.Shared` https://github.com/dotnet/coreclr/pull/20637

File: src/System.Data.SqlClient/src/System/Data/SqlClient/SNI/SNIPacket.cs [hunk 2]
```diff
@@ -21,7 +21,7 @@ internal class SNIPacket : IDisposable, IEquatable<SNIPacket>
         private string _description;
         private SNIAsyncCallback _completionCallback;
 
-        private ArrayPool<byte>  _arrayPool = ArrayPool<byte>.Shared;
+        //private ArrayPool<byte>  _arrayPool = ArrayPool<byte>.Shared;
```
Comment (benaadams) [inline src/System.Data.SqlClient/src/System/Data/SqlClient/SNI/SNIPacket.cs]: Don't need commented out line

File: src/System.Data.SqlClient/src/System/Data/SqlClient/TdsParserStateObject.cs [hunk 3]
```diff
@@ -4043,4 +4137,6 @@ internal void PrepareReplay()
                 }
         */
     }
+
+
```
Comment (benaadams) [inline src/System.Data.SqlClient/src/System/Data/SqlClient/TdsParserStateObject.cs]: Whitespace?

Comment (Wraith2) [reply src/System.Data.SqlClient/src/System/Data/SqlClient/TdsParserStateObject.cs]: gone.

File: src/System.Data.SqlClient/src/System/Data/SqlClient/SNI/SNIPacket.cs (hunk 2 above)
Comment (Wraith2) [reply src/System.Data.SqlClient/src/System/Data/SqlClient/SNI/SNIPacket.cs]: gone.

**Comment by saurabh500** [comment]
oh no not this guy again !!! :smile: 
Thanks for the changes and improving SqlClient. We will review the changes soon.

**Comment by saurabh500** [comment]
Do you have any measurements for the impact of these changes?

File: src/System.Data.SqlClient/src/Interop/SNINativeMethodWrapper.Windows.cs [hunk 4]
```diff
@@ -182,10 +182,10 @@ internal struct SNI_Error
 
         #region DLL Imports
         [DllImport(SNI, CallingConvention = CallingConvention.Cdecl, EntryPoint = "SNIAddProviderWrapper")]
-        internal static extern uint SNIAddProvider(SNIHandle pConn, ProviderEnum ProvNum, [In] ref uint pInfo);
+        internal static extern uint SNIAddProvider(SNISessionHandle pConn, ProviderEnum ProvNum, [In] ref uint pInfo);
```
[the remark this answers is not in the document]
Comment (saurabh500) [reply src/System.Data.SqlClient/src/Interop/SNINativeMethodWrapper.Windows.cs]: I understand that you wanted to make this look clearer but this code is present in .Net Framework and the name is available in the SNI.dll native code as well (which is closed source)
For the sake of keeping the code in sync and not having to wonder about the names, can you revert to the good old SNIHandle ? It makes syncing code easier and different folks work on these parts, I don't want a name change to cause a confusion

[the remark this answers is not in the document]
Comment (Wraith2) [reply src/System.Data.SqlClient/src/Interop/SNINativeMethodWrapper.Windows.cs]: makes sense, done.

**Comment by Wraith2** [comment]
I haven't done any direct measurements because these changes are on a different branch to my other perf changes in #32811. While I was working on the other PR I got to the point where the allocations were all putting values into objects in the parameters (unavoidable) and the IntPtr boxing for the various methods changed here. So logically it should roughly half the GC frequency seen in the graphs posted on the other PR.

Now there are deeper concerns. I'm changing an object which is a single reference in size to a struct 4 times the size which is going to have some impact. When I looked at the uses of the new types all of them were used with a small number of other parameters so while there is more data present now there should be enough registers to allow them to be passed quickly or by reference. Doing some microbenching I could try to identify if it is faster to force references with `in' modifiers but realistically the cpu time is spent waiting on sql server and going to that low level is over optimising at this time.

In performance terms it will be a memory net win. It may have a small throughput degredation but that is likely to be balanced by the lack of boxing instructions and memory operations related to it, it might even be a net win on that side.

**Comment by Wraith2** [comment]
Ok, got some profile results (still can't get BDN to co-operate) they show that there is a minor improvement in both memory and speed. This is a slightly improved version of the benchmark written for 
#32811 where I've removed some overhead from the benchmark setup and stopped using prepared statements. The results are only comparable to each other not between PR's. 100x1000 float inserts.

Before:
before

After:
after

Less GC peaks and probably because of that ~13% faster.

**Comment by Wraith2** [comment]
@dotnet-bot Test Linux x64 Release Build

**Comment by Wraith2** [comment]
EFCore tests passed on this change.

**Comment by Wraith2** [comment]
Nudge.

**Comment by benaadams** [comment]
@dotnet-bot test Windows x64 Debug Build

**Comment by Wraith2** [comment]
For easier reivew and merging I have split this into https://github.com/dotnet/corefx/pull/34040 and https://github.com/dotnet/corefx/pull/34044.

**Pull request closed without merging**